### DISCUSSION TITLE
Prepare edit response on base Element class

### DISCUSSION THREAD

**Discussion by bencroker** [craftcms/cms#10784] [Ideas] [answered]
In Craft 4, sometimes it is desirable to prepare an element's edit page in the CP, for example:

- Register asset bundles
- Set a selected subnav item
- Add alternative actions to the save button
- Make other changes to the response

I am currently dealing with this in my own `edit` controller action, which attaches what it needs to the response from the `elements/edit` controller action before returning it, but it might be nice to be able to put this logic into the base `Element` class. Is there a recommended place for this to go, or could a new method be added, for example:

```php
public function prepEditResponse(Response|CpScreenResponseBehavior $response): void
{
    Craft::$app->getView()->registerAssetBundle(...);
    Craft::$app->getView()->getTwig()->addGlobal('selectedSubnavItem', '...');

    $response->addAltAction(...);
    $response->submitButtonLabel = '...';
}
```

My WIP `edit` controller action:
https://github.com/putyourlightson/craft-campaign/blob/03201b631b5c5db14e3b67709c92548a2e970a85/src/controllers/CampaignsController.php#L70-L110

**Comment by brandonkelly** [comment, marked as answer]
We just released 4.0.0-beta.3 with a new `prepareEditScreen()` method. It takes two arguments:

- `yii\web\Response $response` – The current response being prepared (which will have a CpScreenResponseBehavior attached to it)
- `string $containerId` – The ID that will be assigned to the element editor’s container element

**UPDATE**

The next release will also include the following changes (2329b5acc8b8cf09e35c7936013dcb714277b58a):

- There’s a new `CpScreenResponseBehavior::selectedSubnavItem()` method, which can be called from elements’ `prepareEditScreen()` methods.
- I’ve removed `ElementInterface::getCrumbs()`. Elements can set the crumbs via `CpScreenResponseBehavior::crumbs()` from `prepareEditScreen()` methods as well.

**Reply by bencroker** [reply]
Great stuff!! The only issue is that setting the selected subnav item is not possible here. I suppose I can attach this earlier in the process from within the main plugin class, but is there somewhere specific you would suggest to put this?

```php
public function prepareEditScreen(Response $response, string $containerId): void
{
    Craft::$app->view->getTwig()->addGlobal('selectedSubnavItem', 'campaigns');
}
```
Results in:
```
An Error occurred while handling another error:
LogicException: Unable to add global "selectedSubnavItem" as the runtime or the extensions have already been initialized. in /var/www/project/cms_v4/vendor/twig/twig/src/Environment.php:766
Stack trace:
#0 /Users/ben/Sites/plugins/craft_v4/craft-campaign/src/elements/CampaignElement.php(694): Twig\Environment->addGlobal('selectedSubnavI...', 'campaigns')
#1 /var/www/project/cms_v4/vendor/craftcms/cms/src/controllers/ElementsController.php(723): putyourlightson\campaign\elements\CampaignElement->prepareEditScreen(Object(craft\web\Response), 'main-form')
#2 /var/www/project/cms_v4/vendor/craftcms/cms/src/controllers/ElementsController.php(417): craft\controllers\ElementsController->_prepareEditor(Object(putyourlightson\campaign\elements\CampaignElement), true, Object(craft\web\Response), 'main-form', Object(Closure), Object(Closure), Object(Closure))
#3 [internal function]: craft\controllers\ElementsController->craft\controllers\{closure}(Object(craft\web\Response), 'main-form')
#4 /var/www/project/cms_v4/vendor/craftcms/cms/src/web/CpScreenResponseFormatter.php(107): call_user_func(Object(Closure), Object(craft\web\Response), 'main-form')
#5 /var/www/project/cms_v4/vendor/craftcms/cms/src/web/CpScreenResponseFormatter.php(48): craft\web\CpScreenResponseFormatter->_formatTemplate(Object(craft\web\Response), Object(craft\web\CpScreenResponseBehavior))
#6 /var/www/project/cms_v4/vendor/yiisoft/yii2/web/Response.php(1095): craft\web\CpScreenResponseFormatter->format(Object(craft\web\Response))
#7 /var/www/project/cms_v4/vendor/craftcms/cms/src/web/Response.php(274): yii\web\Response->prepare()
#8 /var/www/project/cms_v4/vendor/yiisoft/yii2/web/Response.php(339): craft\web\Response->prepare()
#9 /var/www/project/cms_v4/vendor/yiisoft/yii2/web/ErrorHandler.php(136): yii\web\Response->send()
#10 /var/www/project/cms_v4/vendor/craftcms/cms/src/web/ErrorHandler.php(184): yii\web\ErrorHandler->renderException(Object(LogicException))
#11 /var/www/project/cms_v4/vendor/yiisoft/yii2/base/ErrorHandler.php(135): craft\web\ErrorHandler->renderException(Object(LogicException))
#12 /var/www/project/cms_v4/vendor/craftcms/cms/src/web/ErrorHandler.php(63): yii\base\ErrorHandler->handleException(Object(LogicException))
#13 [internal function]: craft\web\ErrorHandler->handleException(Object(LogicException))
#14 {main}
```

**Reply by michaelrog** [reply]
Maybe name it `prepareEditScreenResponse()` for clarity?

**Reply by brandonkelly** [reply]
@bencroker Sorry, forgot to address that. Just added a new `selectedSubnavItem()` method to `CpScreenResponseBehavior`, which you can from `prepareEditScreen()`, for the next release. I’ve also removed `Elementinterface::getCrumbs()`, in favor of elements calling `CpScreenResponseBehavior::crumbs()` themselves from `prepareEditScreen()`. (2329b5acc8b8cf09e35c7936013dcb714277b58a)

**Reply by bencroker** [reply]
Perfect, thanks!

**Reply by brandonkelly** [reply]
Craft 4.0.0-beta.4 is out now with those additional changes.
